### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by XxJuliaTruongxX**
Includes workflow Github Actions and FixedTimeSampler feature

### CODE REVIEW

File: smcpy/smc/samplers.py [hunk 1]
```diff
@@ -346,43 +343,30 @@ def sample(
...
-        ):
-            self._phi_linear_slope = (0 - np.log(self._previous_phi)) / (
-                self.final_time - self._previous_step_time
-            )
-            self._buffer_phi = self._previous_phi
+        if not self._buffer_phi and previous_step_time >= self.buffer_time:
+            self._buffer_phi = phi
 
-        self._time_per_step_sequence.append(self._previous_step_time - self._start_time)
+        self._time_per_step_sequence.append(previous_step_time - self._start_time)
```
Comment (peleser) [inline smcpy/smc/samplers.py:351]: "time per step" to me implies that this is the time consumed in the last step _only_. Whereas this is just a trace of the the time relative to the start time. Let's call this `self._time_history` instead.

File: smcpy/smc/samplers.py [hunk 2]
```diff
@@ -346,43 +343,30 @@ def sample(
...
             particles=particles, phi_old=phi_old, target_ess=target_ess
         )
 
-        highest_adaptive = None
-        if self._previous_step_time and self._previous_step_time >= self.buffer_time:
+        if self._time_per_step_sequence[-1] >= self.buffer_time:
             estimated_future_time = (
-                self._time_per_step_sequence[-1] - self._time_per_step_sequence[-2]
-            ) + self._previous_step_time
+                2 * self._time_per_step_sequence[-1] - self._time_per_step_sequence[-2]
```
Comment (peleser) [inline smcpy/smc/samplers.py:360]: Why `2 *` here?

File: tests/unit/test_samplers.py [hunk 3]
```diff
@@ -341,129 +341,128 @@ def test_valid_target_ess(target_ess, mcmc_kernel):
...
 
-    expected_take_loglinear = np.exp(-0.3)
-    expected_take_adaptive = np.exp(-0.1)
+    expected_take_loglinear = 10**-0.3
+    expected_take_adaptive = 10**-0.1
 
     original_phi = smc.optimize_step(particles=1, phi_old=1)
-    smc._do_smc_step(phi=1, num_mcmc_samples=1)
-    assert original_phi == np.exp(-0.9)
+    smc._do_smc_step(phi=original_phi, num_mcmc_samples=1)
```
[the remark this answers is not in the document]
Comment (peleser) [reply tests/unit/test_samplers.py:427]: I realize this comment isn't super clear, so we can talk this one out if you'd like. Let me know.

File: smcpy/smc/samplers.py (hunk 2 above)
Comment (XxJuliaTruongxX) [reply smcpy/smc/samplers.py:360]: This was the change where we replaced `self._previous_step_time` with `self.time_per_step_sequence[-1]. So my thought process was:

old code
```
estimated_future_time = self._time_per_step_sequence[-1] - self._time_per_step_sequence[-2]
            ) **+ self._previous_step_time**
```

replace `self._previous_step_time` with `self.time_per_step_sequence[-1]
```
estimated_future_time = self._time_per_step_sequence[-1] - self._time_per_step_sequence[-2]
            ) + **self._time_per_step_sequence[-1]**
```

since we are adding self._time_per_step_sequence[-1] twice to calculate for estimated_future_time, I just condensed it into **2 * self._time_per_step_sequence[-1]**. Now we have
```
estimated_future_time = 2 * self._time_per_step_sequence - self._time_per_step_sequence[-2]
```

Comment (peleser) [reply smcpy/smc/samplers.py:360]: What, do you have some sort of math minor or something? 😄 Looks good, thanks for clarifying.

File: smcpy/smc/samplers.py [hunk 4]
```diff
@@ -346,43 +343,30 @@ def sample(
...
+            )
 
-            highest_adaptive = max(
-                np.log(curr_adaptive_phi),
+            return 10 ** max(
+                np.log10(curr_adaptive_phi),
                 np.interp(
```
Comment (peleser) [inline smcpy/smc/samplers.py:365]: I think that this code likely covers this corner case, but what happens if `estimated_future_time` is greater than `self.final_time`? Ideally, we just want $\log(\phi)=0$, which I think is what the default extrapolation behavior is, but can you write a test for that and verify?

File: tests/unit/test_samplers.py [hunk 5]
```diff
@@ -338,3 +338,188 @@ def test_valid_target_ess(target_ess, mcmc_kernel):
...
+
+def test_fixed_time_predict_phi_overshoot(mocker, mcmc_kernel):
+    time = 100
+
+    mocker.patch(SAMPLERS + ".AdaptiveSampler.optimize_step")
+    smc = FixedTimeSampler(
+        mcmc_kernel=mcmc_kernel,
+        time=time,
+    )
+    smc._time_history.append(50)
```
Comment (peleser) [inline tests/unit/test_samplers.py:450]: Am I following correctly that this sets the previous time step to 50, and then the next predicted time would be over allotted time? If so, let's check when it's exactly at time as well, even though that should never happen in practice.

File: smcpy/utils/storage.py [hunk 6]
```diff
@@ -122,7 +122,7 @@ def save_step(self, step):
     def _open_h5(self, mode):
         self._refresh_filesystem_metadata()
 
-        h5 = h5py.File(self._filename, mode)
+        h5 = h5py.File(self._filename, mode, track_order=True)
```
[the remark this answers is not in the document]
Comment (peleser) [reply smcpy/utils/storage.py:125]: See new issue re: checking if this is tested and removing redundant code

File: tests/unit/test_samplers.py [hunk 7]
```diff
@@ -338,3 +338,188 @@ def test_valid_target_ess(target_ess, mcmc_kernel):
...
+    )
+
+    smc = FixedTimeSampler(
+        mcmc_kernel=mcmc_kernel,
+        time=time,
+        norm_time_threshold=norm_time_threshold,
+        time_knockdown_factor=time_knockdown_factor,
+    )
+    smc._start_time = 50
+    smc._time_history.append(49)
```
Comment (peleser) [inline tests/unit/test_samplers.py:498]: Is this append required if you're running the smc step with mocked time?

File: tests/unit/test_samplers.py [hunk 8]
```diff
@@ -338,3 +338,188 @@ def test_valid_target_ess(target_ess, mcmc_kernel):
...
+        return_value=0.1,
+    )
+
+    smc = FixedTimeSampler(
+        mcmc_kernel=mcmc_kernel,
+        time=time,
+        norm_time_threshold=norm_time_threshold,
+        time_knockdown_factor=time_knockdown_factor,
+    )
+    smc._start_time = 1
```
Comment (peleser) [inline tests/unit/test_samplers.py:524]: I think this can be removed

File: tests/unit/test_samplers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (XxJuliaTruongxX) [reply tests/unit/test_samplers.py:498]: I ended up deleting this test that tries to test for relative time in the if statements. I took a look at the if statement conditionals in `_do_smc_step` and `optimize_step` and we've made quite a few changes from the original if statements (the buggy ones where I used `time.time()` in the conditionals). Now these conditionals are relying on `_time_history` and the `prev_time_step` (I haven't pushed this change yet so you probably don't see it right now), and I've already tested for those 2 properties. What do you think, is it too bold to say we should get rid of this test?

File: tests/unit/test_samplers.py [hunk 9]
```diff
@@ -439,15 +442,16 @@ def test_fixed_time_predict_buffer_phi(mocker, mcmc_kernel):
     assert smc._buffer_phi == 0.6
 
 
-def test_fixed_time_predict_phi_overshoot(mocker, mcmc_kernel):
+@pytest.mark.parametrize("last_time_history", ((30), (50)))
+def test_fixed_time_predict_phi_overshoot(mocker, mcmc_kernel, last_time_history):
```
Comment (peleser) [inline tests/unit/test_samplers.py:446]: I think a better name might be something like `test_fixted_time_phi_is_one_on_time_overshoot` since phi is not what we are overshooting.

File: tests/unit/test_samplers.py [hunk 10]
```diff
@@ -338,3 +338,189 @@ def test_valid_target_ess(target_ess, mcmc_kernel):
...
+    )
+
+    smc = FixedTimeSampler(
+        mcmc_kernel=mcmc_kernel,
+        time=time,
+    )
+
+    original_phi = smc.optimize_step(particles=1, phi_old=1)
+    numpy_mock.assert_not_called()
+    np.testing.assert_almost_equal(original_phi, 0.1)
```
Comment (peleser) [inline tests/unit/test_samplers.py:526]: I think you should be able to remove the `np.testing` call and just assert that they're equal. If you have a floating point issue you can use `pytest.approx` or just change it to an integer since it's just a mocked value.

File: tests/unit/test_samplers.py [hunk 11]
```diff
@@ -338,3 +338,189 @@ def test_valid_target_ess(target_ess, mcmc_kernel):
...
+    assert smc._buffer_phi == None
+
+    smc._do_smc_step(phi=0.6, num_mcmc_samples=1)
+    assert smc._buffer_phi == 0.6
+
+    smc._do_smc_step(phi=0.9, num_mcmc_samples=1)
+    assert smc._buffer_phi == 0.6
+
+
+def test_fixed_time_predict_buffer_phi(mocker, mcmc_kernel):
```
Comment (peleser) [inline tests/unit/test_samplers.py:419]: I'm still having trouble with this test. I think the intention is to make sure it is interpolating and taking the corrected phi value, since no other test actually checks that? But I'm not positive that's what is happening. If you agree with this intention, let's make it more explicit (see below). If you disagree, let me know and we can talk through it.

To test the interpolation, we need to make sure that our buffer phi has been set to some convenient value. I'd recommend setting up the test such that, before running `optimize_step`, the knockdown factor is 1, the time is 100, buffer phi is 10**-5, the buffer time is 50, the latest time in the time history is 50, and the latest time step is 10 to avoid the "two step" trigger. Then, the next `optimize_step` call should begin the interpolation process, and the interpolation between logphi=-5 and 0 with a timestep of 1/5 the total remaining buffer time should give you logphi=-4 or phi = 10**-4. Verify that and call the test done.
